Implement ssh tunnel reconnection

jeff303 opened the pull request:
From the `connection-with-timezone` method of `execute`, check whether an ssh tunnel that should be open actually is not, and if so, mark the entire pool as invalid (thereby forcing the connection code to rebuild the source and open a new tunnel)

Fixing the `create-pool!` function so that the relevant ssh tunnel entries are kept (in addition to the `:datasource`)

jeff303 committed:
Implement ssh tunnel reconnection

From the connection-with-timezone method of execute, check whether an ssh tunnel that should be open actually is not, and if so, mark the entire pool as invalid (thereby forcing the connection code to rebuild the source and open a new tunnel)

Fixing the create-pool! function so that the relevant ssh tunnel entries are kept (in addition to the :datasource)

Adding test in ssh-test namespace, which will test that the tunnel is reestablished (for now, running with Postgres driver)

jeff303 committed:
Implement ssh tunnel reconnection

From the connection-with-timezone method of execute, check whether an ssh tunnel that should be open actually is not, and if so, mark the entire pool as invalid (thereby forcing the connection code to rebuild the source and open a new tunnel)

Fixing the create-pool! function so that the relevant ssh tunnel entries are kept (in addition to the :datasource)

Adding test in ssh-test namespace, which will test that the tunnel is reestablished (for now, running with Postgres driver)

jeff303 commented:
Looking for feedback on the best place to add a test for this.  All of the existing tests I've found related to ssh tunneling are more narrowly focused (ex: `metabase.util.ssh-test` for the tunnel being opened correctly).  I'm looking for something that goes through the `connection-with-timezone` path with a tunnel, if it exists already (or else I'd need to create that).

Once complete, this should address #10081.

To manually test this functionality (giving specific steps here, but others could obviously be swapped out):

1. Spin up a local MySQL server: `docker run --name some-mysql -P -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:5.7`
2. Figure out the port number that was mapped out of Docker (or use `-p` for stable number)
3. Set up a DB in Metabase, with the MySQL driver, pointing to `localhost` and that port
4. Set up SSH tunneling on the driver (and make sure your host machine allows ssh connections)
5. Run a question against some table and confirm you can see results
6. `ps aux | grep sshd`, and kill the `sshd` process running as the same user that started Metabase
7. `kill` that `sshd` process
8. Reload the question page, and it should work
9. Confirm a log line like the following is printed: `2021-01-28 04:23:16,003 WARN sql-jdbc.connection :: ssh tunnel for database 193 appears to have closed; closing the pool so it can be reestablished`

jeff303 commented on test/metabase/util/ssh_test.clj:
```diff
@@ -198,3 +204,35 @@
+                                     ;; trying to use the mock sshd instance doesn't work
+                                     ;; fails with:
+                                     ;;2021-01-28 23:06:04,956 WARN channel.ClientChannelPendingMessagesQueue :: operationComplete(ClientChannelPendingMessagesQueue[channel=TcpipClientChannel[id=0, recipient=-1]-ClientSessionImpl[jsmith@/127.0.0.1:12221], open=false]) SshChannelOpenException[Generic error while opening channel: 0] signaled
+                                     :tunnel-port        ssh-mock-server-with-password-port
```
@robdaemon , could use your eyes here.  Basically, I can't make this work with the "mock" ssh server that this test starts up.  Not sure what I'm doing wrong.  I'm basically trying to just use the same params as in `connects-with-correct-password`, but it fails to establish the tunnel in the first place.  If I swap these out for my real ssh port and username/password (provided via OS X), then this test works fine.  Perhaps I'm misunderstanding how to actually use the mock fixtures here.

jeff303 replied:
Figured it out.  The local password server needed to have the `AcceptAllForwardingFilter`, like the key server already did.

dpsutton commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -157,6 +165,23 @@
+   remove it from the cache. This will force the next call to db->pooled-connection-spec to reopen the tunnel."
+  [database]
+  (let [conn-spec (db->pooled-connection-spec database)]
+    (if
```
this should be a `when` which has an implicit `do`:

dpsutton replied:
```clojure
(when (and (ssh/use-ssh-tunnel? conn-spec)
               (not (ssh/ssh-tunnel-open? conn-spec)))
      (log/warn
       (u/format-color
           'red
           (trs "ssh tunnel for database {0} appears to have closed; closing the pool so it can be reestablished"
                (u/the-id database))))
      (invalidate-pool-for-db! database))
```

jeff303 replied:
Fixed.

jeff303 committed:
Responding to PR feedback from Dan

jeff303 committed:
Responding to PR feedback from Dan

jeff303 committed:
Fixing test by adding AcceptAllForwardingFilter forwardingFilter to the mock password server instance

jeff303 committed:
Fixing test by adding AcceptAllForwardingFilter forwardingFilter to the mock password server instance

jeff303 committed:
Change with-driver to test-driver

jeff303 commented on src/metabase/driver/sql_jdbc/execute.clj:
```diff
@@ -151,6 +151,7 @@
 
 (defmethod connection-with-timezone :sql-jdbc
   [driver database ^String timezone-id]
+  (sql-jdbc.conn/invalidate-pool-if-ssh-tunnel-closed! database)
```
Unclear why this isn't covered...

jeff303 replied:
OK, thanks to the new `test-ssh-tunnel-reconnection-h2` test, this is now covered even for H2, and hence for the coverage check.

jeff303 replied:
This is finally covered, after a series of changes.

jeff303 committed:
Add ssh tunnel reconnect test that can run against H2

jeff303 committed:
Change test-driver to with-driver in hopes of making CodeCov finally happy

jeff303 committed:
Adding new multimethod to ssh namespace, called incorporate-ssh-tunnel-details, for accounting for the

Implementing incorporate-ssh-tunnel-details for h2 so update the URI string (:db key) to point to the tunnel entry point

Pulled logic for :sql-jdbc implementation of connection-with-timezone out to a new fn, so it can be called elsewhere

Updating H2 reconnection test imn light of the changes above

Added "!" suffix to name of include-ssh-tunnel to reflect the fact that it does modify global state

jeff303 committed:
Fixing NPE in incorporate-ssh-tunnel-details implementation for :h2

Moving multimethod declaration for incorporate-ssh-tunnel-details to driver namespace

jeff303 committed:
Fix :h2 implementation again

jeff303 committed:
Remove another errant extra line :(

dpsutton commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -166,7 +188,7 @@
   "Return an appropriate JDBC connection spec to test whether a set of connection details is valid (i.e., implementing
   `can-connect?`)."
   [driver details]
-  (let [details-with-tunnel (ssh/include-ssh-tunnel details)]
```
this logic seems off to me. the `include-ssh-tunnel` and `incorporate-ssh-tunnel-details` each attempt to start an ssh tunnel. Seems like this is no good? If the tunnel is already open the `can-connect?` function will blow up on the tunnel's port already open?

jeff303 replied:
Yeah, although this point could have been made about the existing code as well.  In any case, it's updated in the latest review to only go through the "open" logic if required.

dpsutton commented on src/metabase/util/ssh.clj:
```diff
@@ -126,16 +132,27 @@
       details-with-tunnel)
     details))
 
+(defmethod driver/incorporate-ssh-tunnel-details nil
```
why would a driver ever be nil? I don't think this could or should happen

jeff303 replied:
Some test cases for databases (either in `api` or `models`) had such a situation.  Rather than trying to go down the rabbit hole of rewriting them (I think they were using some contrived test engines that somehow didn't fully register), this seemed to be the easier fix.

jeff303 replied:
Actually, seems unnecessary (removed).

dpsutton commented:
I was expecting to see something like the following in `db->pooled-connection-spec`:
```clojure
       (locking database-id->connection-pool
         (or
          ;; check if another thread created the pool while we were waiting to acquire the lock
          (when-let [details (get @database-id->connection-pool database-id)]
            (when (or (nil? (:tunnel-session details))
                      (.isOpen ^ClientSession (:tunnel-session details))) ;; just check that either there is no tunnel, or if there is, its open. 
              details)) ;; and if its not open, we just pretend we didn't see it and we will fall through to creating a new pool underneath here
          ;; create a new pool and add it to our cache, then return it
          (let [db (or (db/select-one [Database :id :engine :details] :id database-id)
                       (throw (ex-info (tru "Database {0} does not exist." database-id)
                                       {:status-code 404
                                        :type        qp.error-type/invalid-query
                                        :database-id database-id})))]
            (u/prog1 (create-pool! db)
              (set-pool! database-id <>)))))
```

This seems like it would be a good choke point for anyone that uses a connection to ensure the tunnel is open. And then the only thing remaining would be to do a bit of surgery for all of the `can-connect?` methods. that way you don't have to remember to call the `invalidate-if-closed-function` since you won't get a connection that has a closed tunnel anyways.

jeff303 committed:
Rebase onto master

jeff303 committed:
Rebase again to fix merge conflict

dpsutton commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -166,7 +190,7 @@
   `can-connect?`)."
   [driver details]
-  (let [details-with-tunnel (ssh/include-ssh-tunnel details)]
+  (let [details-with-tunnel (driver/incorporate-ssh-tunnel-details driver details)]
```
this `incorporate-ssh-tunnel-details` being a multimethod seems strange to me. As it works right now, all sql based stuff gets it, but not mongo and others. The logic doesn't seem like it would change on the type of driver: if there's a tunnel if it's open we're good, otherwise re-open it. that seems fundamental and not to depend on the type of database.

dpsutton replied:
also, its name gives me no indication that it will actually reopen an ssh tunnel.

robdaemon replied:
The logic might change per driver - some drivers are RESTful APIs instead of JDBC.

dpsutton replied:
how would that change a metabase managed ssh connection though?

jeff303 replied:
Refer to the H2 implementation.  That's why I made it a multimethod, so that it could be supported in H2.  Most of the JDBC DBs use `:host` and `:port` separately, but H2 actually just has `:db` which contains the full connection string.  When the tunnel is opened, the port portion of that URL has to be swapped out dynamically for the tunnel entry port.

dpsutton replied:
I see. I read that on this a while back and assumed it was just validation of tcp rather than seeing it hot-swapping the port. I think maybe the version you have for `:sql-jdbc` should be moved to the driver namespace as `:default` so everything gets that except for h2. Don't see any reason its just sql-jdbc stuff.

jeff303 replied:
I think my rationale was that the `driver.clj` file seemed to be fairly sparse, and everything that used the tunnel previously had `:sql-jdbc` as a parent.  Moving the default implementation there would end up importing the `ssh` namespace and a bunch of stuff there.  But if that makes more sense, I'm happy to make the change.

dpsutton replied:
what else would be imported? For sure ssh but it makes sense for the driver stuff to depend on ssh. But i think it makes the most sense for all connections to get this behavior unless they have a particular way to do it (like h2)

jeff303 replied:
Hmm, actually it seems this introduces a cyclic dependency.

`Cyclic load dependency: [ /metabase/driver ]->/metabase/util/ssh->[ /metabase/driver ]->/metabase/plugins/lazy_loaded_driver->/metabase/plugins/initialize->/metabase/plugins`

Might be a mess to untangle. 😢

dpsutton replied:
i don't think so. this change added the metabase.driver dep into ssh. so just move it and restart the repl and you should be good to go.

jeff303 replied:
Good point.  Backed that out, but different error of the same kind.

```
Syntax error compiling at (metabase/driver/util.clj:1:1).
Cyclic load dependency: [ /metabase/driver ]->/metabase/driver/util->/metabase/public_settings->/metabase/util/ssh->[ /metabase/driver ]->/metabase/plugins/lazy_loaded_driver->/metabase/plugins/initialize->/metabase/plugins
```

dpsutton replied:
i think that's still a remnant of your broken state. it's saying that m.u.ssh depends on driver which is no longer true right?

jeff303 replied:
I don't think so, now.  Looking on `master`...

- `ssh` depends on `public-settings`
- `public-settings` depends on `driver.util`
- `driver.util` depends on `driver`

Therefore, we can't add a dependency on `ssh` to `driver` or else we end up with this cycle.

dpsutton replied:
ahh. i didn't have that change that added the dependency on public-settings. i'm sorry about that. I guess to leave this as `:sql-jdbc`. Looking a bit further, its only called from the jdbc stuff anyways so we can do it this way for now. I apologize for not reading closer @jeff303 !

dpsutton commented on src/metabase/driver/sql_jdbc/execute.clj:
```diff
@@ -168,6 +168,10 @@
 
+(defmethod connection-with-timezone :sql-jdbc
+  [driver database ^String timezone-id]
+  (default-connection-with-timezone driver database timezone-id))
```
why was this extracted into a function?

jeff303 replied:
To be honest, I have no idea anymore.  I'll back this out.

jeff303 replied:
I suspect originally it was so the H2 implementation could call it as well, but that's not needed in the latest changes.

dpsutton commented on src/metabase/util/ssh.clj:
```diff
@@ -129,27 +136,46 @@
       details-with-tunnel)
     details))
 
+;(defmethod driver/incorporate-ssh-tunnel-details nil
```
dont think we need this commented out code any longer

jeff303 committed:
Inline the private helper fn back into connection-with-timezone

jeff303 committed:
Remove dead code
